Google Books

adam3smith opened the pull request:
Account for new format, simplify

adam3smith commented on Google Books.js:
```diff
@@ -2,16 +2,38 @@
 	"translatorID": "3e684d82-73a3-9a34-095f-19b112d88bbf",
 	"label": "Google Books",
 	"creator": "Simon Kornblith, Michael Berkowitz and Rintze Zelle",
```
Still need to add myself here

adam3smith commented on Google Books.js:
```diff
@@ -2,16 +2,38 @@
+/*
+	***** BEGIN LICENSE BLOCK *****
+
+	Copyright © 2019 Sebastian Karcher
```
And add other contributors here

zuphilip commented:
I guess that this then fixes #2049.

adam3smith commented on Google Books.js:
```diff
@@ -58,62 +61,74 @@ https://play.google.com/store/books/details/Adam_Smith_The_Wealth_of_Nations?id=
+	var items = {};
+	var found = false;
+	// TODO: adjust the CSS selector
+	var rows = doc.querySelectorAll('div.bHexk>a, div.Q9MA7b>a');
```
yeah, the first one works for all regular searches, the 2nd one for all google play ones. Will add a comment to explain. I know the class names look wonky, but I looked through quite a number of different versions & searches and it worked everywhere. The alternative would be a longer `div` tree without attribute, which I felt was more fragile.

zuphilip replied:
When I open the mentioned search page in my browser, then I don't see any `div`s of class `bHexk`. Do you? Is this page depending on the regions where you are located?

![firefox_2019-11-10_12-13-51](https://user-images.githubusercontent.com/5199995/68543154-b3df3300-03b3-11ea-898f-2e125549cc2a.jpg)

adam3smith replied:
oh wow, yes, everything below `<div class="srg">` looks completely different for me. 
I think I'll go with `ZU.xpath(doc, '//div[@class="srg"]//a[h3])` and then just test for length and handle the play store separately.
![image](https://user-images.githubusercontent.com/128986/68556352-f3337100-03ff-11ea-9c9c-78551ce86864.png)

adam3smith commented on Google Books.js:
```diff
@@ -58,62 +61,74 @@ https://play.google.com/store/books/details/Adam_Smith_The_Wealth_of_Nations?id=
+	for (let row of rows) {
+		let href = row.href;
+		// TODO: check and maybe adjust
+		let title = ZU.trimInternal(text(row, 'h3, div'));
```
done

adam3smith commented on Google Books.js:
```diff
@@ -58,62 +61,74 @@ https://play.google.com/store/books/details/Adam_Smith_The_Wealth_of_Nations?id=
+	var rows = doc.querySelectorAll('div.bHexk>a, div.Q9MA7b>a');
+	for (let row of rows) {
+		let href = row.href;
+		// TODO: check and maybe adjust
```
done

adam3smith commented on Google Books.js:
```diff
@@ -58,62 +61,74 @@ https://play.google.com/store/books/details/Adam_Smith_The_Wealth_of_Nations?id=
+	var id;
+	// New books format:
+	if (url.includes("/books/edition/")) {
+		id = url.match(/\/books\/edition.*\/([^/]+)(?:\?.+)?$/)[1];
```
The regex captures from the last slash to ? or the end of the URL -- both are possible. Not sure I see the difference?

zuphilip replied:
The regex itself does IMO not say to take the last slash, because more slashes could be captured afterwards in the `.+` operator, but one could also make this more explicit by using `[^/]+` instead. (Or does greediness of the operator makes this unique anyway? But not sure we should rely on that...)

adam3smith replied:
Yeah, this relies on the greediness of `.*` but your version is simpler so going with that

adam3smith commented:
Thanks -- fixed or responded. Yeah, I looked for a solution to the author issue, but they're not just in the XML data but also in the page and the bibtex Google offer for the new google books layout, so I really wouldn't know how to avoid them.

zuphilip commented:
@adam3smith Let me know when you could look into the open problem https://github.com/zotero/translators/pull/2052#discussion_r344459684  and fix it.

adam3smith commented:
@zuphilip please take another look. Could you check if the playstore works for you / looks the same for you? I'm worried that class also changes depending on locale/user account/googles current mood

zuphilip commented on Google Books.js:
```diff
@@ -1,17 +1,39 @@
-	"creator": "Simon Kornblith, Michael Berkowitz and Rintze Zelle",
-	"target": "^https?://(books|www)\\.google\\.[a-z]+(\\.[a-z]+)?/(books(/.*)?\\?(.*id=.*|.*q=.*)|search\\?.*?(btnG=Search\\+Books|tbm=bks))|^https?://play\\.google\\.[a-z]+(\\.[a-z]+)?/(store/)?(books|search\\?.+&c=books)",
+	"creator": "Simon Kornblith, Michael Berkowitz, Rintze Zelle, and Sebastian Karcher",
+	"target": "^https?://(books|www)\\.google\\.[a-z]+/(books(/.*)?\\?(.*id=.*|.*q=.*)|search\\?.*?(btnG=Search\\+Books|tbm=bks)|books/edition/)|^https?://play\\.google\\.[a-z]+(\\.[a-z]+)?/(store/)?(books|search\\?.+&c=books)",
```
What about audiobooks, e.g. https://play.google.com/store/audiobooks/details/Arthur_Conan_Doyle_Der_Hund_von_Baskerville?id=AQAAAECM-EG43M which is also linked in the test case below?

zuphilip replied:
Moreover, also the following link is possible https://play.google.com/store/search?c=books&q=kling but not recognized.

adam3smith commented:
@zuphilip if you'd take another look?
I've added support for  https://play.google.com/store/search?c=books&q=kling type URLs in the play store and removed detection/selection for audiobooks -- their metadata isn't included in the standard google book metadata API and I think it's fine to not cover them. The play store is probably a fairly marginal source here anyway

zuphilip commented:
Great! Thank you @adam3smith ! This is now squashed and merged :rocket:
